test/e2e/storage/api/snapshotresource.go

```diff
@@ -0,0 +1,372 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package api
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/onsi/ginkgo"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/types"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/uuid"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/kubernetes/test/e2e/framework"
+	"k8s.io/kubernetes/test/e2e/storage/utils"
+)
+
+// SnapshotResource represents a snapshot class, a snapshot and its bound snapshot contents for a specific test case
+type SnapshotResource struct {
+	Config  *PerTestConfig
+	Pattern TestPattern
+
+	Vs        *unstructured.Unstructured
+	Vscontent *unstructured.Unstructured
+	Vsclass   *unstructured.Unstructured
+}
+
+// CreateSnapshot creates a VolumeSnapshotClass with given SnapshotDeletionPolicy and a VolumeSnapshot
+// from the VolumeSnapshotClass using a dynamic client.
+// Returns the unstructured VolumeSnapshotClass and VolumeSnapshot objects.
+func CreateSnapshot(sDriver SnapshottableTestDriver, config *PerTestConfig, pattern TestPattern, pvcName string, pvcNamespace string, timeouts *framework.TimeoutContext) (*unstructured.Unstructured, *unstructured.Unstructured) {
```

Comment:
To me, this seems more of a util compared to the testsuite framework because it's helpers for specific tests. What is your method for distinction?

Reply:
I think you are right, I need to move this to util

Reply:
I doubled check on this and find it is not easy to do so because this function as you can tell from the signature: 
func CreateSnapshot(sDriver SnapshottableTestDriver, config *PerTestConfig, pattern TestPattern, pvcName string, pvcNamespace string, timeouts *framework.TimeoutContext)

It has a lot of dependency of storage framework(SnapshottableTestDriver, PerTestConfig, TestPattern). That being said, it cannot be moved to storageutils. From my understanding, storageutils should be self-encapsulated. The dependency direction is framework can depend on storageutils while storageutils cannot depend on framework...

I feel like there are a lot of code refactor we can do around the utils for the testsuites. But I think this can be our first step towards that. Let's make this in and then we can babystep the utils refactor and cleanup slowly.